### PR TITLE
[WIP]BIGTOP-3760. Bump Hadoop to 3.3.4.

### PR DESCRIPTION
**Pull request by iwasakims**
https://issues.apache.org/jira/browse/BIGTOP-3760

Hadoop 3.3.4 RC0 is under vote.
https://lists.apache.org/thread/r41ftyrs4b6g67sf8nz8xypfwohov7qb

### CODE REVIEW

**Comment by iwasakims** [comment]
smoke-tests of hdfs,yarn,mapreduce passed on Rocky Linux 8 x86_64:

```
$ ./docker-hadoop.sh \
     --create 3 \
     --image bigtop/puppet:trunk-rockylinux-8 \
     --memory 8g \
     --repo file:///bigtop-home/output \
     --disable-gpg-check \
     --stack hdfs,yarn,mapreduce \
     --smoke-tests hdfs,yarn,mapreduce
...
BUILD SUCCESSFUL in 13m 14s
41 actionable tasks: 14 executed, 27 up-to-date
```

**Comment by iwasakims** [comment]
Building HBase 2.4.12 against Hadoop 3.3.4 failed. The cause seems to be added dependencies. Fixing exclusion rule in HBase side might be the fix.

```
[ERROR] Found artifact with unexpected contents: '/home/rocky/srcs/bigtop/build/hbase/rpm/BUILD/hbase-2.4.12/hbase-shaded/hbase-shaded-mapreduce/target/hbase-shaded-mapreduce-2.4.12.jar'
    Please check the following and either correct the build or update
    the allowed list with reasoning.

    okhttp3/
    okhttp3/Request$Builder.class
    ...
    kotlin/
    kotlin/collections/
    kotlin/collections/ArraysUtilJVM.class
    ...
```

**Comment by iwasakims** [comment]
Building Hive 3.1.3 and Spark 3.2.1 against Hadoop 3.3.4 worked. smoke-tests passed.

```
$ ./docker-hadoop.sh \
     --create 3 \
     --image bigtop/puppet:trunk-rockylinux-8 \
     --memory 8g \
     --repo file:///bigtop-home/output \
     --disable-gpg-check \
     --stack hdfs,yarn,mapreduce,hive,spark \
     --smoke-tests hive,spark
...
Now testing...
:bigtop-tests:smoke-tests:spark:test (Thread[Execution worker for ':' Thread 7,5,main]) completed. Took 1 mins 11.596 secs.

BUILD SUCCESSFUL in 2m 0s
36 actionable tasks: 11 executed, 25 up-to-date
Stopped 1 worker daemon(s).
```

**Comment by iwasakims** [comment]
Building HBase succeded with added patch1-relocate-okhttp.diff. smoke-tests passed.

```
$ ./docker-hadoop.sh \
     --create 3 \
     --image bigtop/puppet:trunk-rockylinux-8 \
     --memory 8g \
     --repo file:///bigtop-home/output \
     --disable-gpg-check \
     --stack hdfs,yarn,mapreduce,hbase \
     --smoke-tests hbase
...
ow testing...
:bigtop-tests:smoke-tests:hbase:test (Thread[Execution worker for ':' Thread 3,5,main]) completed. Took 3 mins 50.955 secs.

BUILD SUCCESSFUL in 5m 10s
36 actionable tasks: 10 executed, 26 up-to-date
```

**Comment by brahmareddybattula** [comment]
@iwasakims thanks for working on this. any chance test this sqoop and Tez also..? and RC-0 will be re-place once the release is announced ..?

**Comment by iwasakims** [comment]
We can not build Tez until #963 is merged. I'm not willing to test all downstream products in this PR. Please test it after merging this PR then file a JIRA if you find issues. The download URL will be updated after 3.3.4 release then [WIP] will be removed from the title. @brahmareddybattula

**Comment by brahmareddybattula** [comment]
> We can not build Tez until #963 is merged. I'm not willing to test all downstream products in this PR. Please test it after merging this PR then file a JIRA if you find issues. 
I assumed all the downstream will be verified when we introduced the major version.  Since sqoop and tez is not mentioned I asked. Saw the TEZ #963 which is blocking. Once TEZ PR merged, will check.. thanks.
>The download URL will be updated after 3.3.4 release then [WIP] will be removed from the title. @brahmareddybattula
ok, thanks.

**Comment by guyuqi** [comment]
Let me test it on AArch64, thanks for working on it.

**Comment by guyuqi** [comment]
> Building HBase succeded with added patch1-relocate-okhttp.diff. smoke-tests passed.
> 
> ```
> $ ./docker-hadoop.sh \
>      --create 3 \
>      --image bigtop/puppet:trunk-rockylinux-8 \
>      --memory 8g \
>      --repo file:///bigtop-home/output \
>      --disable-gpg-check \
>      --stack hdfs,yarn,mapreduce,hbase \
>      --smoke-tests hbase
> ...
> ow testing...
> :bigtop-tests:smoke-tests:hbase:test (Thread[Execution worker for ':' Thread 3,5,main]) completed. Took 3 mins 50.955 secs.
> 
> BUILD SUCCESSFUL in 5m 10s
> 36 actionable tasks: 10 executed, 26 up-to-date
> ```

Tested it on aarch64 Rocky Linux8 :
HBase: **2.4.13**
```
[root@2862226bc183 /]# hbase version
OpenJDK 64-Bit Server VM warning: Using incremental CMS is deprecated and will likely be removed in a future release
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/lib/hbase/lib/client-facing-thirdparty/slf4j-reload4j-1.7.33.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/lib/hadoop/lib/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Reload4jLoggerFactory]
HBase 2.4.13
```

Hadoop: **3.3.4**:
```
[root@2862226bc183 /]# hadoop version
Hadoop 3.3.4
Source code repository https://github.com/apache/bigtop.git -r 22a0813b12f244bb9ba3df05532505dd7fed73e7
Compiled by root on 2022-07-25T06:29Z
Compiled with protoc 3.7.1
From source with checksum fb9dd8918a7b8a5b430d61af858f6ec
This command was run using /usr/lib/hadoop/hadoop-common-3.3.4.jar

```



```
Gradle Test Executor 5 finished executing tests.

> Task :bigtop-tests:smoke-tests:hbase:test
Finished generating test XML results (0.039 secs) into: /bigtop-home/bigtop-tests/smoke-tests/hbase/build/test-results/test
Generating HTML test report...
Finished generating test html results (0.052 secs) into: /bigtop-home/bigtop-tests/smoke-tests/hbase/build/reports/tests/test
Now testing...
:bigtop-tests:smoke-tests:hbase:test (Thread[Execution worker for ':' Thread 72,5,main]) completed. Took 7 mins 47.886 secs.

BUILD SUCCESSFUL in 32m 21s

```

what about to remove `patch1-relocate-okhttp.diff` patched for HBase `2.4.12`?

**Comment by iwasakims** [comment]
> what about to remove patch1-relocate-okhttp.diff which was just only for HBase 2.4.12?

I updated the patch1 based on HBase 2.4.13. We still need the fix for HBase 2.4.13. @guyuqi

**Comment by guyuqi** [comment]
It failed to build Hbase (on Rocklinux and Debian 11)  for x86_64/aarch64.

```
..........
...
[INFO] Apache HBase - Common .............................. FAILURE [ 40.535 s]
..........
....
[ERROR] Failed to execute goal on project hbase-common: Could not resolve dependencies for project org.apache.hbase:hbase-common:jar:2.4.13: Could not find artifact org.apache.hadoop:hadoop-common:jar:3.3.4 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
```

**Comment by iwasakims** [comment]
```
[ERROR] Failed to execute goal on project hbase-common: Could not resolve dependencies for project org.apache.hbase:hbase-common:jar:2.4.13: Could not find artifact org.apache.hadoop:hadoop-common:jar:3.3.4 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
```

You need to build Hadoop first (by `./gradlew hadoop-pkg`) to install artifacts of Hadoop 3.3.4 to local repository. It is not found in public Maven repositories since it is not released yet.

**Comment by iwasakims** [comment]
> On the other hand I successfully tested smoke tests of Hadoop-3.3.4 and Hbase 2.4.13 without patch1-relocate-okhttp.diff.

The patch is for **building** HBase against Hadoop 3.3.4. I guess you used HBase 2.4.13 already built against Hadoop 3.3.3 here.

**Comment by guyuqi** [comment]
> ```
> [ERROR] Failed to execute goal on project hbase-common: Could not resolve dependencies for project org.apache.hbase:hbase-common:jar:2.4.13: Could not find artifact org.apache.hadoop:hadoop-common:jar:3.3.4 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
> ```
> 
> You need to build Hadoop first (by `./gradlew hadoop-pkg`) to install artifacts of Hadoop 3.3.4 to local repository. It is not found in public Maven repositories since it is not released yet.

ok,  let me try to build Hadoop-3.3.4 and Hbase-2.4.13 in the single docker container.
`docker run --rm -v `pwd`:/ws --workdir /ws bigtop/slaves:trunk-rockylinux-8 bash -c '. /etc/profile.d/bigtop.sh; ./gradlew hadoop-clean hbase-clean hadoop-pkg hbase-pkg'`

**Comment by guyuqi** [comment]
+1.

Tested smoke tests of `Spark` and `HBase` with `Hadoop-3.3.4` on Rocklinux8 and Debian-11 for both x86 and Arm64.
Hive smoke tests were also passed on x86_64 but it failed to build Hive on Arm64. I'd like to fix it tracked by [BIGTOP-3763](https://issues.apache.org/jira/browse/BIGTOP-3763).

**Comment by iwasakims** [comment]
https://lists.apache.org/thread/y2779v6bw88q9lt25pqfc7fhlf4pz95k

**Comment by iwasakims** [comment]
Packaging and smoke-tests of Hadoop, HBase, Hive and Spark passed as expected on RC1.

File: bigtop.bom [hunk 1]
```diff
@@ -145,11 +145,11 @@ bigtop {
     'hadoop' {
       name    = 'hadoop'
       relNotes = 'Apache Hadoop'
-      version { base = '3.3.3'; pkg = base; release = 1 }
+      version { base = '3.3.4'; pkg = base; release = 1 }
       tarball { destination = "${name}-${version.base}.tar.gz"
                 source      = "${name}-${version.base}-src.tar.gz" }
-      url     { download_path = "/$name/common/$name-${version.base}"
-                site = "${apache.APACHE_MIRROR}/${download_path}"
+      url     { download_path = "/$name/$name-${version.base}-RC1"
```
Comment (guyuqi) [inline bigtop.bom:151]: Thanks for updating it. @iwasakims 
May I ask when would Hadoop upstream guys make the stable release for 3.3.4, or they prefer to offer release candidate 2 (RC2)?

Comment (iwasakims) [reply bigtop.bom:151]: @guyuqi No one is willing to cut another RC before new issue is found on ongoing vote. There is no veto now.

**Comment by guyuqi** [comment]
+1
Verified Spark, hive, hbase, Zeppelin smoke tests against on this Hadoop-3.3.4-RC1 on Arm64/x86_64 rockylinux-8 and Debian-11.

**Comment by iwasakims** [comment]
@guyuqi We need to update `download_path` and `site` to make it point to released artifact. (It is the reason why the title is prefixed with **[WIP]** now..)

I will add addendum commit for this. The vote passed. The artifacts will be published soon.

**Comment by guyuqi** [comment]
> @guyuqi No one is willing to cut another RC before new issue is found on ongoing vote. There is no veto now.

I'm sorry I misunderstand about "There is no vote now" and I don't konw when the stable version 3.3.4 would release.
As you mentioned that the 3.3.4 vote was recently passed,  please add the addendum commit after the 3.3.4 published, thanks.

**Comment by iwasakims** [comment]
17fc6eaa9a33241b4d3a7e917b110201d54a8931 is the addendum commit.

**Pull request merged**